src/pocketmine/tile/Banner.php

```diff
@@ -0,0 +1,198 @@
+	const PATTERN_MOJANG = "moj";
+
+	private $patternCount = 0;
+
+	public function __construct(Level $level, CompoundTag $nbt){
+		if(!isset($nbt->Base) or !($nbt->Base instanceof IntTag)){
+			$nbt->Base = new IntTag("Base", 15);
+		}
+		if(!isset($nbt->Patterns) or !($nbt->Patterns instanceof ListTag)){
+			$nbt->Patterns = new ListTag("Patterns");
+		}
+		parent::__construct($level, $nbt);
+	}
+
+	public function addAdditionalSpawnData(CompoundTag $nbt){
+		$nbt->Patterns = $this->namedtag->Patterns;
+		$nbt->Base = $this->namedtag->Base;
+	}
+
+	/**
+	 * Returns the color of the banner base.
+	 *
+	 * @return int
+	 */
+	public function getBaseColor() : int{
+		return $this->namedtag->Base->getValue();
+	}
+
+	/**
+	 * Sets the color of the banner base.
+	 *
+	 * @param int $color
+	 */
+	public function setBaseColor(int $color){
+		$this->namedtag->Base->setValue($color & 0x0f);
+		$this->onChanged();
+	}
+
+	/**
+	 * Applies a new pattern on the banner with the given color.
+	 *
+	 * @param string $pattern
+	 * @param int    $color
+	 *
+	 * @return int ID of pattern.
+	 */
+	public function addPattern(string $pattern, int $color) : int{
+		$this->namedtag->Patterns->{($id = $this->patternCount++)} = new CompoundTag("", [
+			new IntTag("Color", $color & 0x0f),
+			new StringTag("Pattern", $pattern)
+		]);
+		$this->onChanged();
+		return $id;
+	}
+
+	/**
+	 * Changes the pattern of a previously existing pattern.
+	 *
+	 * @param int    $id
+	 * @param string $pattern
+	 * @param int    $color
+	 *
+	 * @return bool indicating success.
+	 */
+	public function changePattern(int $id, string $pattern, int $color) : bool{
+		if(!isset($this->namedtag->Patterns->{$id})){
+			return false;
+		}
+		$this->namedtag->Patterns->{$id}->setValue([
+			new IntTag("Color", $color & 0x0f),
+			new StringTag("Pattern", $pattern)
+		]);
+		$this->onChanged();
+		return true;
+	}
+
+	/**
+	 * Deletes a layer of the banner with the given ID.
+	 *
+	 * @param int $patternLayer
+	 *
+	 * @return bool indicating whether the pattern existed or not.
+	 */
+	public function deletePattern(int $patternLayer) : bool{
+		if(!isset($this->namedtag->Patterns->{$patternLayer})){
+			return false;
+		}
+		unset($this->namedtag->Patterns->{$patternLayer});
+		return true;
+	}
+
+	/**
+	 * Deletes the top most pattern of the banner.
+	 *
+	 * @return bool indicating whether the banner was empty or not.
+	 */
+	public function deleteTopPattern() : bool{
+		if(empty((array) $this->namedtag->Patterns)){
+			return false;
+		}
+		$index = (int) max(array_keys((array) $this->namedtag->Patterns));
+		unset($this->namedtag->Patterns->{$index});
+		return true;
+	}
+
+	/**
+	 * Deletes the bottom pattern of the banner.
+	 *
+	 * @return bool indicating whether the banner was empty or not.
+	 */
+	public function deleteBottomPattern() : bool{
+		if(empty((array) $this->namedtag->Patterns)){
+			return false;
+		}
+		$index = (int) min(array_keys((array) $this->namedtag->Patterns));
```

Comment:
Just gone through the code, I don't think the (int) casting is needed.

Reply:
Sometimes I would get "\0" here, which would lead to an error. I'm not entirely sure of the reason, but this fixed it.

Reply:
Note that all strings starting with non-numeric bytes, including non-printable bytes ("\0" and "\1"), are converted to int(0) when cast to int. Make sure you don't make everything zero.

Reply:
Properly fixed now.